chainerx_cc/chainerx/cuda/cuda_device/linalg.cu

```diff
@@ -0,0 +1,208 @@
+#include "chainerx/cuda/cuda_device.h"
+
+#include <cstdint>
+#include <mutex>
+#include <type_traits>
+
+#include <cublas_v2.h>
+#include <cuda_runtime.h>
+#include <cusolverDn.h>
+#include <cuda_fp16.hpp>
+
+#include "chainerx/array.h"
+#include "chainerx/axes.h"
+#include "chainerx/backend.h"
+#include "chainerx/backend_util.h"
+#include "chainerx/cuda/cublas.h"
+#include "chainerx/cuda/cuda_runtime.h"
+#include "chainerx/cuda/cuda_set_device_scope.h"
+#include "chainerx/cuda/cusolver.h"
+#include "chainerx/cuda/data_type.cuh"
+#include "chainerx/cuda/float16.cuh"
+#include "chainerx/cuda/kernel_regist.h"
+#include "chainerx/device.h"
+#include "chainerx/dtype.h"
+#include "chainerx/error.h"
+#include "chainerx/float16.h"
+#include "chainerx/kernels/creation.h"
+#include "chainerx/kernels/linalg.h"
+#include "chainerx/kernels/misc.h"
+#include "chainerx/macro.h"
+#include "chainerx/native/native_device.h"
+#include "chainerx/routines/creation.h"
+#include "chainerx/routines/linalg.h"
+
+namespace chainerx {
+namespace cuda {
+namespace {
+
+template <typename T>
+cusolverStatus_t GetrfBuffersize(cusolverDnHandle_t /*handle*/, int /*m*/, int /*n*/, T* /*a*/, int /*lda*/, int* /*lwork*/) {
+    throw DtypeError{"Only Arrays of float or double type are supported by getrf (LU)"};
+}
+
+template <typename T>
+cusolverStatus_t Getrf(
+        cusolverDnHandle_t /*handle*/, int /*m*/, int /*n*/, T* /*a*/, int /*lda*/, T* /*workspace*/, int* /*devipiv*/, int* /*devinfo*/) {
+    throw DtypeError{"Only Arrays of float or double type are supported by getrf (LU)"};
+}
+
+template <typename T>
+cusolverStatus_t Getrs(
+        cusolverDnHandle_t /*handle*/,
+        cublasOperation_t /*trans*/,
+        int /*n*/,
+        int /*nrhs*/,
+        T* /*a*/,
+        int /*lda*/,
+        int* /*devipiv*/,
+        T* /*b*/,
+        int /*ldb*/,
+        int* /*devinfo*/) {
+    throw DtypeError{"Only Arrays of float or double type are supported by getrs (Solve)"};
+}
+
+template <>
+cusolverStatus_t GetrfBuffersize<double>(cusolverDnHandle_t handle, int m, int n, double* a, int lda, int* lwork) {
+    return cusolverDnDgetrf_bufferSize(handle, m, n, a, lda, lwork);
+}
+
+template <>
+cusolverStatus_t GetrfBuffersize<float>(cusolverDnHandle_t handle, int m, int n, float* a, int lda, int* lwork) {
+    return cusolverDnSgetrf_bufferSize(handle, m, n, a, lda, lwork);
+}
+
+template <>
+cusolverStatus_t Getrf<double>(cusolverDnHandle_t handle, int m, int n, double* a, int lda, double* workspace, int* devipiv, int* devinfo) {
+    return cusolverDnDgetrf(handle, m, n, a, lda, workspace, devipiv, devinfo);
+}
+
+template <>
+cusolverStatus_t Getrf<float>(cusolverDnHandle_t handle, int m, int n, float* a, int lda, float* workspace, int* devipiv, int* devinfo) {
+    return cusolverDnSgetrf(handle, m, n, a, lda, workspace, devipiv, devinfo);
+}
```

Comment:
Can these functions (above 3) be implemented without templates with regular overloads if it's the matter of three overloads?

Reply:
Yes.
My initial intent was to raise `DtypeError` for types other than implemented double and float with these templates.
I guess it can be changed to overloads and make an explicit check for dtype.

Reply:
It was decided during the meeting that we'll keep the current approach with the templated functions.